Run Docker container as unprivileged user, allow PUID/PGID selection

lkubb opened the pull request:
Currently, the container process runs with root privileges. This cannot be changed, even by specifying PUID/PGID (as suggested in `docker-compose.yml`).

This PR migrates to running as an unprivileged user and makes it possible to specify PUID/PGID environment variables to choose UID/GID. Migration of existing data owned by root is handled transparently.

A side-effect is that the /app directory is read-only to the python process, I'm not sure if that is a problem since I'm not familiar with the architecture of changedetection.

I also took the liberty to clean up the Dockerfile (and apt cache) a bit, hope that's not a problem. :) Please feel free to suggest changes or deny the PR, if it does not fit your vision.

Fixes https://github.com/dgtlmoon/changedetection.io/issues/565.

lkubb committed:
Run Docker container as unprivileged user

Previously, the container process ran with root privileges.
This commit changes the default to an unprivileged user and
makes it possible to specify PUID/PGID environment variables
to choose UID/GID.
Migration of existing data owned by root is handled transparently.

lkubb committed:
Improve docker entrypoint to work rootless

* The previous behavior required root privileges on start, now this is
optional (when run in rootless mode with --user). The user has to ensure
the correct ownership is set for existing data in that case.

* Furthermore, it is now possible to choose to keep root privileges by
setting KEEP_PRIVILEGES to something non-empty.

* This also renames abc user to changedetection (nicer).

dgtlmoon commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,27 @@
+
+set -eu
+
+# If the first argument looks like a flag, assume we want to run changedetection
```
hmmm what is the problem you're trying to solve here?

lkubb replied:
Assuming you're referring to the line below: If the container is e.g. run with `docker-compose run changedetection -c`, this would run `python ./changedetection.py -d datastore -c` inside the container. It's mostly a convention, nothing that's necessary, just nice to have. If you prefer it removed, no worries. :)

dgtlmoon commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,27 @@
+    usermod -o -u "$PUID" changedetection
+
+    # Look for files in datadir not owned by the correct user and chown them
+    find "/datastore" \! -user changedetection -exec chown changedetection '{}' +
```
I'm wondering if there could be a case where it would NOT be `/datastore` ? and if we can auto-detect what that path would be?

dgtlmoon replied:
Also any reference to `/datastore` should be in a single var if possible `$DATSTORE_DIR` ?

lkubb replied:
It definitely should be defined in a var, agreed.

Off the top of my head, I can't think of a reason to override the defaults for the `/datadir` inside the container since anything can be mounted there. Autodetection might add unnecessary complexity + some unforeseen quirks imho, so I would assume anyone actually needing this for some reason should be experienced enough to do the necessary adjustments by themselves. We could parse the CLI arguments of course, but I'm not sure what the benefit would be.

dgtlmoon reviewed (commented):
see PR comments - mainly about being sure it's actually `/datastore` ?

dgtlmoon commented on Dockerfile:
```diff
@@ -33,20 +33,24 @@ FROM python:3.8-slim
 # https://stackoverflow.com/questions/58701233/docker-logs-erroneously-appears-empty-until-container-stops
 ENV PYTHONUNBUFFERED=1
 
-RUN [ ! -d "/datastore" ] && mkdir /datastore
```
hmmm should we keep this? :)

lkubb committed:
Use variable for datastore in Docker entrypoint

dgtlmoon commented on Dockerfile:
```diff
@@ -33,34 +33,44 @@ FROM python:3.8-slim
+        libxslt-dev \
+        zlib1g-dev && \
+    rm -rf /var/lib/apt/lists/*; \
+    useradd -u 911 -U -d /datastore -s /bin/false changedetection && \
```
why `911` ? :) Is it better to choose a UID > 1000 ?

lkubb replied:
It's the standard one used by [linuxserver.io](https://github.com/linuxserver) and makes it a system user, but a higher default one like 3000 is fine as well imho. It can be chosen freely anyways (docker-entrypoint.sh:L16-20).

lkubb commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,29 @@
+# If we're running as root, by default make sure process uid/gid
+# and datadir permissions are correct. This can be skipped by setting
+# KEEP_PRIVILEGES to something non-empty.
+if [ "$(id -u)" = '0' -a -n "${KEEP_PRIVILEGES:-}" ]; then
```
This is the wrong way around though, should be `-a -z` to work as intended.

lkubb committed:
Correct semantics on KEEP_PRIVILEGES in Docker entrypoint

dgtlmoon commented:
`master` now tests the application inside the fully built docker container

also need to test this on an existing install which used root.root

dgtlmoon commented on Dockerfile:
```diff
@@ -62,4 +71,5 @@ COPY changedetection.py /app/changedetection.py
 
-CMD [ "python", "./changedetection.py" , "-d", "/datastore"]
+ENTRYPOINT ["/app/docker-entrypoint.sh"]
+CMD ["python", "./changedetection.py", "-d", "/datastore"]
```
You define `/datastore` at line 58 in a env var here and then never use it?

lkubb replied:
Good catch, although `never` is wrong (see entrypoint script). I probably wanted to avoid using a shell, hence no env vars are accessible. Will come up with a solution.

dgtlmoon commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,29 @@
+    find "$DATASTORE_DIR" \! -user changedetection -exec chown changedetection '{}' +
+
+    # Restart this script as an unprivileged user
+    exec gosu changedetection:changedetection "$BASH_SOURCE" "$@"
```
goso wasnt installed, kills the container

lkubb replied:
That's because you deleted the line with your merge, it was there... https://github.com/dgtlmoon/changedetection.io/pull/722/commits/48565082d4da01872cacae1c1fc0bb4346115fc3

dgtlmoon replied:
@lkubb oops thanks :pray: could you add it back and update the branch with master? :pray:

lkubb replied:
Sure :)

dgtlmoon replied:
Thank you! then we can see if it will build across all architectures

lkubb replied:
Just saw you also deleted the user account generation. We need that to run rootless. :) I merged master, let's see how that works.

lkubb replied:
Now I missed a ``\`` in the ``RUN`` part, sorry. I force-pushed the correction.

dgtlmoon reviewed (commented):
unsure

dgtlmoon commented:
@lkubb biggest one so far, gosu not found

dgtlmoon commented:
hmm when it's running as test, it needs to use `/datastore` but only when run in github/automated, because often people test locally when developing too

dgtlmoon commented:
https://github.com/dgtlmoon/changedetection.io/blob/e99f07a51d84b351f6dbeda0e04d62596a66fd96/changedetectionio/tests/conftest.py#L33

somehow needs a var passed to `pytest` like `pytest -d /datastore` or.. unsure

lkubb commented:
a) Introduce a new env var `DATASTORE_TESTING`, which defaults to `test-datastore`, but override it in the github workflow 
b) Just `mkdir` the test directory in the docker build script. Not very clean but it shouldn't hurt.

dgtlmoon commented:
> b) Just `mkdir` the test directory in the docker build script. Not very clean but it shouldn't hurt.

This is fine if theres a comment above like
`# creating test-dir for pytest to run in`

lkubb committed:
Create test datastore for pytest to run in

lkubb commented:
I added `mkdir` earlier, but you will need to approve the workflow.

dgtlmoon commented on Dockerfile:
```diff
@@ -31,35 +31,49 @@ RUN pip install --target=/dependencies playwright~=1.27.1 \
 # The eventlet server wrapper
 COPY changedetection.py /app/changedetection.py
 
+RUN chmod 777 /app/docker-entrypoint.sh && \
```
why are we doing all this work then letting it be run and edited by anyone? `777` ?

lkubb replied:
I have no answer to that. `755` is sufficient.

dgtlmoon commented:
/app/changedetectionio/.pytest_cache/

but why not just `chown` all the `/app` to the same UID/GUID that we run the container as?

lkubb commented:
Because usually, there is no good reason for an application to be able to modify itself. We could of course disable all that with a build arg specifically for the workflow, but I would be very hesitant to `chown` the app folder for production.

lkubb committed:
Correct mode on entrypoint script

lkubb committed:
Do not write bytecode in Docker container

dgtlmoon commented:
haha I love tests, what looked like a simple PR always shows up something

```
FAILED tests/test_backup.py::test_backup - PermissionError: [Errno 13] Permis...
```

so `/datastore` directory isnt writable (it creates the backup ZIP there)

lkubb commented:
>  so /datastore directory isnt writable (it creates the backup ZIP there)

From what I can tell, it creates it as `/app/changedetectionio/download.zip`, right?

```console
============================= test session starts ==============================
platform linux -- Python 3.10.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /app/changedetectionio, configfile: pytest.ini
plugins: flask-1.2.0
collected 1 item

tests/test_backup.py::test_backup 
[...]
>       with open("download.zip", 'wb') as f:
E       PermissionError: [Errno 13] Permission denied: 'download.zip'

dgtlmoon commented:
Oops right you are, that test needs fixing so its using the `test-datastore` directory

dgtlmoon commented:
Or better is that it doesnt even write the file to disk and just stores it memory

lkubb committed:
Fix datastore path handling

dgtlmoon commented:
please change

```
    # ZipFile from buffer seems non-obvious, just save it instead
    with open("download.zip", 'wb') as f:
        f.write(res.data)

    zip = ZipFile('download.zip')
```

to

```
    from io import BytesIO, StringIO
    zip = ZipFile(BytesIO(res.data))
```

lkubb commented:
Already did, try running the workflow again.

dgtlmoon commented:
cool, and what about people who have `/datastore` mounted as root already (like everyone :) ) ? i guess that will continue to work, needs testing somehow

lkubb commented:
That should definitely work since it's handled by the entrypoint script with `chown -R`. Honestly, imho a manual test should be sufficient since it's a simple and one-time operation inside a very specific environment. I'm not too eager to create an automated test for that inside the current testing infrastructure.

On a side-note regarding the tests in general:

For `test-datastore`, you might consider using something like this instead:

```python
@pytest.fixture(scope='session')
def datastore_path(tmp_path_factory):
    yield tmp_path_factory.mktemp("test-datastore")
```

This will make sure the directory is writable an clean up after itself. This would also allow to remove the `mkdir -p /app/changedetectionio/test-datastore && chown -R changedetection:changedetection /app/changedetectionio/test-datastore` from the Dockerfile.

Is there a reason each `test/test_*.py` is run separately? Pytest autodiscovers those files by itself, so instead of

```bash
find tests/test_*py -type f|while read test_name
do
  echo "TEST RUNNING $test_name"
  pytest $test_name
done
```
```bash
pytest tests/
```
should be enough.

dgtlmoon commented on Dockerfile:
```diff
@@ -31,35 +31,51 @@ RUN pip install --target=/dependencies playwright~=1.27.1 \
+    PYTHONUNBUFFERED=1 \
+    # https://stackoverflow.com/questions/64808915/should-pycache-folders-be-included-in-production-containers
+    # This avoids permission denied errors because the app directory is root-owned.
+    PYTHONDONTWRITEBYTECODE=1 \
```
isnt this going to slow down the app? or not?

re pytest cache, this is maybe better (set it to `datadir` ) https://docs.pytest.org/en/latest/reference/reference.html#confval-cache_dir

lkubb replied:
> The advantage of cached bytecode is in startup time, not execution speed; import takes longer, but actual execution is just as fast (the bytecode is still there in memory, just not on disk). But in the case of containers, making the container larger means it takes longer to cache and launch, so you're wasting disk space and any gains in Python launch time are typically offset by delays pulling and launching the container in the first place. And writing out bytecode at runtime is pointless; it'll be lost when the container shuts down, so it's adding expense to startup for no benefit.

dgtlmoon replied:
ok thanks, yeah that makes sense

alpe12 commented:
Any update on this? @dgtlmoon 
Thanks.

dgtlmoon commented:
> Is there a reason each test/test_*.py is run separately? Pytest autodiscovers those files by itself, so instead of

It seems to fail, can you provide a PR for this @lkubb ?

dgtlmoon commented:
I'de love to pick this one up again in the next few months!

dgtlmoon commented:
Also good to add https://docs.kernel.org/userspace-api/no_new_privs.html to docker-compose.yml

```
    security_opt:
      - no-new-privileges:true
```

dgtlmoon commented on changedetectionio/changedetection.py:
```diff
@@ -32,10 +32,12 @@ def main():
     port = os.environ.get('PORT') or 5000
     do_cleanup = False
-    datastore_path = None
+    datastore_path = os.environ.get('DATASTORE_PATH')
```
not needed anymore

dgtlmoon commented on changedetectionio/changedetection.py:
```diff
@@ -32,10 +32,12 @@ def main():
+        pass
     # On Windows, create and use a default path.
-    if os.name == 'nt':
+    elif os.name == 'nt':
```
not needed anymore

dgtlmoon commented on Dockerfile:
```diff
@@ -31,35 +31,51 @@ RUN pip install --target=/dependencies playwright~=1.27.1 \
+    rm -rf /var/lib/apt/lists/*; \
+    useradd -u 911 -U -M -s /bin/false changedetection && \
+    usermod -G users changedetection; \
+    mkdir -p /datastore
```
mkdir if not exists? will it work if it exists?

duketuxem replied:
`mkdir -p` prevents the command to fail upon existing directory, so the exit code will be 0.

dgtlmoon commented:
https://github.com/docker-library/python/issues/20

lkubb committed:
Use COPY --chmod

lkubb commented:
@duketuxem 

> This PR looked very promising—any chance we might see it merged soon?

Honestly, I got tired of explaining at some point and decided to run the container in Podman rootless with userns remapping.

I rebased on master once again, let's see if the tests still work (@dgtlmoon needs to approve them though). It has been a while since I created this PR, hope I didn't break anything during the rebase.

lkubb commented:
@dgtlmoon It's this fixture:

```py
@pytest.fixture(scope='function')
def measure_memory_usage(request):
    memory_usage = {"peak": 0, "stop": False}
    tracker_thread = Thread(target=track_memory, args=(memory_usage,))
    tracker_thread.start()

    yield

    memory_usage["stop"] = True
    tracker_thread.join()

    # Note: ru_maxrss is in kilobytes on Unix-based systems
    max_memory_used = memory_usage["peak"] / 1024  # Convert to MB
    s = f"Peak memory used by the test {request.node.fspath} - '{request.node.name}': {max_memory_used:.2f} MB"
    logger.debug(s)

    with open("test-memory.log", 'a') as f:
        f.write(f"{s}\n")
```

I can't find a reference to `test-memory.log` in the whole repository, is this still in use? Can it be replaced with a log message?

dgtlmoon commented:
to be honest the whole memory test thing is not needed.. you can remove it, or on another PR, it didnt really add anything interesting :)

lkubb committed:
Don't write to the app directory in tests

lkubb commented:
Alright, I chose to just remove writing the file since the rest of the code does not interfere.

lkubb commented:
> sometimes there are "random" fails, i'll keep an eye on it

👍 It's still complaining that `/app/changedetectionio/.pytest_cache` is not writable, might need to account for that somehow, maybe create the directory in the entrypoint if it detects we're running pytest.

dgtlmoon commented:
hmm https://stackoverflow.com/questions/33310615/is-there-a-way-to-change-the-location-of-pytests-cache-directory could help

lkubb committed:
Don't create Pytest cache in container by default

lkubb committed:
Ensure changedetection has a home dir

... to be able to install extra packages via pip

lkubb commented:
Alright, all tests are passing now. For most setups, the entrypoint script should automatically chown the datastore to the correct user. The one exception is in case it is a network share mount, where it might fail. In this case, admins must ensure the correct `PUID`/`PGID` are set in the container so the script does not attempt to chown any files.

In case this logic causes any problems, it can be skipped completely by setting the `KEEP_PRIVILEGES` environment variable to a non-empty value. This results in the previous behavior.

I'd really like to wrap this PR up soon since it has been going for nearly two and a half years. :)

dgtlmoon commented:
can you think of some test cases to try? like i already have a datastore owned by some user and...

lkubb commented:
First, I would verify everything works as before when `KEEP_PRIVILEGES` is set to provide an immediate bandaid to any problems.

Then, with a root-owned (populated) local datastore:

1) Starting the container changes the ownership to 911:911 and everything works fine.
2) Starting the container with PUID=1000 changes the ownership to 1000:911
3) Starting the container with PGID=1000 changes the ownership to 911:1000
4) Starting the container with PGID=1000 PGID=1000 changes the ownership to 1000:1000

Then chown -R the datastore to 1234:5678 and do 1 and 4 again.

Then chown only select files to a different user/group and see if 1/4) works.

You should check that running the container rootless (still?) works with 1+4.

Tbh I'm not familiar with the application architecture, so I'm not sure if there are any other paths that should be tested.

dgtlmoon commented:
> I'm currently considering if it would be a better idea to just check if a specific uid:gid has write permissions for the data dir and if not, chown -R it once.

better than doing it every time, some datasets are pretty big

lkubb committed:
Don't force the DATASTORE_PATH

This got lost during a rebase, it's not necessary anymore since the env
var is set inside the container by default and changedetection respects it.

lkubb committed:
Only chown if datastore is not writable by PUID:PGID

lkubb commented:
Alright, the entrypoint now checks writability of the datastore by the designated PUID:PGID combination and only intervenes if that fails. I'll adjust the proposed test cases above.

dgtlmoon commented on changedetectionio/__init__.py:
```diff
@@ -42,15 +42,17 @@ def main():
+        pass
     # On Windows, create and use a default path.
-    if os.name == 'nt':
+    elif os.name == 'nt':
```
shouldnt this stay as `if` ?

lkubb replied:
No, it would then not respect the `DATASTORE_PATH` defined by the user (or the default one set during container build time).

dgtlmoon commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,41 @@
+# If we're running as root, by default make sure process uid/gid
+# and datadir permissions are correct. This can be skipped by setting
+# KEEP_PRIVILEGES to something non-empty.
+if [ "$(id -u)" = '0' -a -z "${KEEP_PRIVILEGES:-}" ]; then
```
Maybe all this should only run linux?

The commands groupmod, usermod, and gosu are linux only

lkubb replied:
The container is always a Linux one.

dgtlmoon replied:
ahhh ofcourse :) thanks

leleogere commented:
Thanks for this work! It seems that all tests pass, there are no conflicts with the base branch, the questions have been answered (at least from what I see from the last discussions), and still this PR hasn't been updated for the past 5 months. Is there anything preventing it to be merged?

kimpenhaus commented:
@dgtlmoon is there any chance to get this pr merged? it looks like it is ready to go - and while the current container is running as root and writing files as root I am not able to backup the files running borg backup as unprivileged user:

```
/srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/29a24202d0194baec6c01affcfd8c350.txt.br: open: [Errno 13] Permission denied: '29a24202d0194baec6c01affcfd8c350.txt.br'
E /srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/29a24202d0194baec6c01affcfd8c350.txt.br
/srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/90f37966f5feb05c818dac3873001005.txt.br: open: [Errno 13] Permission denied: '90f37966f5feb05c818dac3873001005.txt.br'
E /srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/90f37966f5feb05c818dac3873001005.txt.br
/srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/a2c1f9154902cd93b7d9cbdc1857bd91.txt.br: open: [Errno 13] Permission denied: 'a2c1f9154902cd93b7d9cbdc1857bd91.txt.br'
E /srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/a2c1f9154902cd93b7d9cbdc1857bd91.txt.br
/srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/6356681e7375d1b9a075783be56afdf4.txt.br: open: [Errno 13] Permission denied: '6356681e7375d1b9a075783be56afdf4.txt.br'
E /srv/nfs/docker/changedetection/16ddaa48-20e3-4e01-afd6-7cbace31a96e/6356681e7375d1b9a075783be56afdf4.txt.br
```
I have high hopes that this PR could solve my issue.

thanks
